### PR TITLE
make startbuild command external

### PR DESCRIPTION
**Pull request by juanvallejo**
Switches the startbuild command to use external versions.
Switches env utils to use external versions.

~~Depends on https://github.com/openshift/origin/pull/20399~~

@deads2k @mfojtik

### CODE REVIEW

File: pkg/pod/envresolve/env.go [hunk 1]
```diff
@@ -71,6 +72,7 @@ func getResourceFieldRef(from *kapi.EnvVarSource, c *kapi.Container) (string, er
 }
 
 // GenEnvVarRefValue returns the value referenced by the supplied EnvVarSource given the other supplied information
+// TODO: how to handle this? It is also used by the build controller which deals with internal versions
```
Comment (juanvallejo) [inline pkg/pod/envresolve/env.go]: @mfojtik 

The set/env command needs this helper to receive a corev1.EnvVarSource and a corev1.Container.
The build controller depends on a separate helper in pkg/build/controller/common which depends on the current (internal) signature of this helper

File: pkg/oc/cli/startbuild/startbuild.go [hunk 2]
```diff
@@ -890,7 +888,7 @@ func gitRefInfo(repo git.Repository, dir, ref string) (buildapi.GitRefInfo, erro
 }
 
 // WaitForBuildComplete waits for a build identified by the name to complete
-func WaitForBuildComplete(c buildclient.BuildResourceInterface, name string) error {
+func WaitForBuildComplete(c buildclientv1.BuildResourceInterface, name string) error {
```
Comment (juanvallejo) [inline pkg/oc/cli/startbuild/startbuild.go]: the versioned / v1 build client does not actually have a `BuildResourceInterface`

**Comment by juanvallejo** [comment]
cc @soltysh

File: pkg/oc/cli/startbuild/startbuild.go [hunk 3]
```diff
@@ -289,24 +288,24 @@ func (o *StartBuildOptions) Complete(f kcmdutil.Factory, cmd *cobra.Command, cmd
...
 
-	o.Namespace = namespace
 	o.Name = name
 
-	o.BuildLogClient = buildclientinternalmanual.NewBuildLogClient(o.BuildClient.RESTClient(), o.Namespace)
+	o.BuildLogClient = buildclientmanual.NewBuildLogClient(o.BuildClient.RESTClient(), o.Namespace)
 
 	// Handle environment variables
 	cmdutil.WarnAboutCommaSeparation(o.ErrOut, o.Env, "--env")
+	// TODO: not sure what to do here, duplicate this helper to return v1 EnvVars?
```
Comment (juanvallejo) [inline pkg/oc/cli/startbuild/startbuild.go]: @deads2k @soltysh

File: pkg/oc/cli/startbuild/startbuild.go [hunk 4]
```diff
@@ -289,24 +288,24 @@ func (o *StartBuildOptions) Complete(f kcmdutil.Factory, cmd *cobra.Command, cmd
...
 	env, _, err := utilenv.ParseEnv(o.Env, o.In)
```
Comment (juanvallejo) [inline pkg/oc/cli/startbuild/startbuild.go]: @deads2k @soltysh

File: pkg/oc/cli/startbuild/startbuild.go [hunk 5]
```diff
@@ -289,29 +288,30 @@ func (o *StartBuildOptions) Complete(f kcmdutil.Factory, cmd *cobra.Command, cmd
...
 	}
 	if len(buildLogLevel) > 0 {
-		env = append(env, kapi.EnvVar{Name: "BUILD_LOGLEVEL", Value: buildLogLevel})
+		env = append(env, corev1.EnvVar{Name: "BUILD_LOGLEVEL", Value: buildLogLevel})
 	}
 	o.EnvVar = env
 
 	// Handle Docker build arguments. In order to leverage existing logic, we
 	// first create an EnvVar array, then convert it to []docker.BuildArg
+	// TODO: also not sure what to do here, duplicate this helper to return v1 EnvVars?
```
Comment (juanvallejo) [inline pkg/oc/cli/startbuild/startbuild.go]: cc @deads2k @soltysh

File: pkg/oc/cli/startbuild/startbuild.go (hunk 4 above)
Comment (soltysh) [reply pkg/oc/cli/startbuild/startbuild.go]: Do in-place conversion using `kcmdutil.AsDefaultVersionedOrOriginal` for now and leave there a TODO. In a followup to this and #20400 you'll sweep those at once.

File: pkg/oc/cli/startbuild/startbuild.go [hunk 6]
```diff
@@ -17,31 +17,31 @@ import (
 	"time"
 
 	"github.com/golang/glog"
+	"github.com/openshift/origin/pkg/oc/util/ocscheme"
```
Comment (soltysh) [inline pkg/oc/cli/startbuild/startbuild.go]: Move down to origin

File: pkg/oc/cli/startbuild/startbuild.go [hunk 7]
```diff
@@ -290,34 +300,55 @@ func (o *StartBuildOptions) Complete(f kcmdutil.Factory, cmd *cobra.Command, cmd
...
 	// Handle Docker build arguments. In order to leverage existing logic, we
 	// first create an EnvVar array, then convert it to []docker.BuildArg
 	buildArgs, err := utilenv.ParseBuildArg(o.Args, o.In)
 	if err != nil {
 		return err
 	}
-	o.BuildArgs = buildArgs
+
+	// convert internal buildargs returned from the helper to external versions
+	externalBuildArgs := []corev1.EnvVar{}
```
Comment (soltysh) [inline pkg/oc/cli/startbuild/startbuild.go]: Make a helper and use here and couple lines above.

File: pkg/oc/cli/startbuild/startbuild.go [hunk 8]
```diff
@@ -828,13 +861,13 @@ func (o *StartBuildOptions) RunStartBuildWebHook() error {
 	body, _ := ioutil.ReadAll(resp.Body)
 	if len(body) > 0 {
 		// In later server versions we return the created Build in the body.
-		var newBuild buildapi.Build
-		if err = json.Unmarshal(body, &buildapiv1.Build{}); err == nil {
-			if err = runtime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), body, &newBuild); err != nil {
+		newBuild := &buildv1.Build{}
+		if err = json.Unmarshal(body, newBuild); err == nil {
```
Comment (soltysh) [inline pkg/oc/cli/startbuild/startbuild.go]: This looks overly complicated, why not doing:
```go
newBuild := &buildv1.Build{}
if err := json.Unmarshal(body, newBuild); err != nil {
    return err
}                
if err := o.Printer.PrintObj(newBuild, o.Out); err != nil {
    return err
}
```

File: pkg/build/client/v1/logs.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package v1
```
Comment (soltysh) [inline pkg/build/client/v1/logs.go]: Shims? From my comparison it looks like only the parameter differs, so in-place conversions to internals and use that, at least for now.

Comment (deads2k) [reply pkg/build/client/v1/logs.go]: This doesn't satisfy a complexity requirement.  A duplicate here is ok.

File: pkg/oc/cli/startbuild/startbuild.go [hunk 10]
```diff
@@ -22,26 +22,27 @@ import (
 	"github.com/openshift/source-to-image/pkg/tar"
 	s2ifs "github.com/openshift/source-to-image/pkg/util/fs"
 
+	"github.com/openshift/origin/pkg/oc/util/ocscheme"
```
Comment (soltysh) [inline pkg/oc/cli/startbuild/startbuild.go]: Move down to origin's imports, besides that scheme should be from k8s, right?

**Pull request merged**